source/extensions/filters/http/admission_control/admission_control.cc

```diff
@@ -0,0 +1,129 @@
+#include "extensions/filters/http/admission_control/admission_control.h"
+
+#include <chrono>
+#include <cstdint>
+#include <string>
+#include <vector>
+
+#include "envoy/extensions/filters/http/admission_control/v3alpha/admission_control.pb.h"
+#include "envoy/runtime/runtime.h"
+#include "envoy/server/filter_config.h"
+
+#include "common/common/assert.h"
+#include "common/common/cleanup.h"
+#include "common/common/enum_to_int.h"
+#include "common/http/utility.h"
+#include "common/protobuf/utility.h"
+
+#include "extensions/filters/http/well_known_names.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace AdmissionControl {
+
+static constexpr double defaultAggression = 2.0;
+static constexpr std::chrono::seconds defaultHistoryGranularity{1};
+
+AdmissionControlFilterConfig::AdmissionControlFilterConfig(
+    const AdmissionControlProto& proto_config, Runtime::Loader& runtime, TimeSource& time_source,
+    Runtime::RandomGenerator& random, Stats::Scope& scope, ThreadLocal::SlotPtr&& tls)
+    : runtime_(runtime), time_source_(time_source), random_(random), scope_(scope),
+      tls_(std::move(tls)), admission_control_feature_(proto_config.enabled(), runtime_),
+      aggression_(
+          proto_config.has_aggression_coefficient()
+              ? std::make_unique<Runtime::Double>(proto_config.aggression_coefficient(), runtime_)
+              : nullptr) {}
+
+double AdmissionControlFilterConfig::aggression() const {
+  return std::max<double>(1.0, aggression_ ? aggression_->value() : defaultAggression);
+}
+
+AdmissionControlFilter::AdmissionControlFilter(AdmissionControlFilterConfigSharedPtr config,
+                                               const std::string& stats_prefix)
+    : config_(std::move(config)), stats_(generateStats(config_->scope(), stats_prefix)) {}
+
+Http::FilterHeadersStatus AdmissionControlFilter::decodeHeaders(Http::RequestHeaderMap&,
+                                                                bool end_stream) {
+  if (!end_stream || !config_->filterEnabled() || decoder_callbacks_->streamInfo().healthCheck()) {
+    return Http::FilterHeadersStatus::Continue;
+  }
+
+  if (shouldRejectRequest()) {
+    decoder_callbacks_->sendLocalReply(Http::Code::ServiceUnavailable, "", nullptr, absl::nullopt,
+                                       "denied by admission control");
+    stats_.rq_rejected_.inc();
+    return Http::FilterHeadersStatus::StopIteration;
+  }
+
+  deferred_record_failure_ =
+      std::make_unique<Cleanup>([this]() { config_->getController().recordFailure(); });
+
+  return Http::FilterHeadersStatus::Continue;
+}
+
+Http::FilterHeadersStatus AdmissionControlFilter::encodeHeaders(Http::ResponseHeaderMap& headers,
+                                                                bool end_stream) {
+  if (end_stream) {
+    const uint64_t status_code = Http::Utility::getResponseStatus(headers);
+    if (status_code < 500) {
```

Comment:
Here would be the point at which I would recommend having a success criteria interface, and then a default implementation.